rpcs3/rpcs3qt/pkg_install_dialog.cpp

```diff
@@ -0,0 +1,132 @@
+#include <QListWidget>
+#include <QToolButton>
+
+constexpr int FullPathRole = Qt::UserRole+0;
+constexpr int BaseDisplayRole = Qt::UserRole+1;
+
+pkg_install_dialog::pkg_install_dialog(std::function<bool(const QString&, bool)> installCb, const QStringList& paths, QWidget* parent)
+	: QDialog(parent)
+{
+	QListWidget* dirList = new QListWidget(this);
+
+	class numbered_widget_item final : public QListWidgetItem
+	{
+	public:
+		explicit numbered_widget_item(const QString& text, QListWidget* listview = nullptr, int type = Type)
+			: QListWidgetItem(text, listview, type)
+		{
+		}
+
+		QVariant data(int role) const override
+		{
+			QVariant result;
+			switch (role)
+			{
+			case Qt::DisplayRole:
+				result = QStringLiteral("%1. %2").arg(listWidget()->row(this) + 1).arg(data(BaseDisplayRole).toString());
+				break;
+			case BaseDisplayRole:
+				result = QListWidgetItem::data(Qt::DisplayRole);
+				break;
+			default:
+				result = QListWidgetItem::data(role);
+				break;
+			}
+			return result;
+		}
+
+		bool operator<(const QListWidgetItem& other) const override
+		{
+			return data(BaseDisplayRole).toString() < other.data(BaseDisplayRole).toString();
+		}
+	};
+
+	for (const QString& path : paths)
+	{
+		QListWidgetItem* item = new numbered_widget_item(QFileInfo(path).fileName(), dirList);
+		// Save full path in a custom data role
+		item->setData(FullPathRole, path);
+	}
+
+	dirList->sortItems();
+	dirList->setCurrentRow(0);
+	dirList->setMinimumWidth((dirList->sizeHintForColumn(0) * 125) / 100);
+
+	// Create buttons
+	QDialogButtonBox* buttons = new QDialogButtonBox(QDialogButtonBox::Cancel | QDialogButtonBox::Ok);
+	buttons->button(QDialogButtonBox::Ok)->setText(tr("Install"));
+	buttons->button(QDialogButtonBox::Ok)->setDefault(true);
+
+	connect(buttons, &QDialogButtonBox::clicked, [this, buttons, dirList, cb = std::move(installCb)](QAbstractButton* button)
+	{
+		if (button == buttons->button(QDialogButtonBox::Ok))
+		{
+			accept();
+
+			for (int i = 0, count = dirList->count(); i < count; i++)
+			{
+				QVariant data = dirList->item(i)->data(FullPathRole);
+				if (!cb(data.toString(), i == (count-1)))
+				{
+					break;
+				}
+			}
+		}
+		else if (button == buttons->button(QDialogButtonBox::Cancel))
+		{
+			reject();
+		}
+	});
+
+	QToolButton* moveUp = new QToolButton;
+	moveUp->setArrowType(Qt::UpArrow);
+	moveUp->setToolTip(tr("Move selected item up"));
+	connect(moveUp, &QToolButton::clicked, [dirList]()
+	{
+		int currentIndex = dirList->currentRow();
+		if (currentIndex > 0)
+		{
+			QListWidgetItem* item = dirList->takeItem(currentIndex);
+			dirList->insertItem(currentIndex - 1, item);
+			dirList->setCurrentItem(item);
+		}
+	});
+
+	QToolButton* moveDown = new QToolButton;
+	moveDown->setArrowType(Qt::DownArrow);
+	moveDown->setToolTip(tr("Move selected item down"));
+	connect(moveDown, &QToolButton::clicked, [dirList]()
+	{
+		int currentIndex = dirList->currentRow();
+		if (currentIndex < dirList->count() - 1)
```

Comment:
Missing check for > 0

Reply:
It's moving down, so moving from `current_index = 0` is valid, and this code will never be executed if `dirList->count()` is less than 2.

Reply:
Doesn't matter. Checking for negative index is programming 101

Reply:
In which case I can probably wire both to the same lambda with 1/-1 as a parameter.

Reply:
Factorized it into a method so now this code is shared.